### DISCUSSION TITLE
Encountering an unexpected exception [java.lang.IllegalArgumentException] when ordering vertices/edges by property values that are not defined for all

### DISCUSSION THREAD

**Discussion by joyemang33** [JanusGraph/janusgraph#3840] [Q&A]
From [here](https://lists.lfaidata.foundation/g/janusgraph-users/topic/bug_encountering_an/97599862?p=,,,20,0,0,0::recentpostdate/sticky,,,20,2,0,97599862,previd%3D1685535865846050219,nextid%3D1671724710195794476&previd=1685535865846050219&nextid=1671724710195794476), and many thanks for reproducing it by @li-boxuan.

```
g.addV("v1")
g.addV("v2").property("name", "joye")
g.V().order().by("name")
```
This will lead to an unexpected exception ```java.lang.IllegalArgumentException```, whereas TinkerGraph will filter out the nodes that miss the key properties.

Notably, if all nodes have no key properties, such exceptions will not be raised.

It would be highly appreciated if you further investigate it and fix it that make the case more consistent.

Best regards,
Joye

**Comment by porunov** [comment]
@joyemang33 this is expected behavior. You can't sort by incomparable values (in your situation a missing value). 
Would you expect `v1` be placed before `v2` or after `v2`? JanusGraph can't decide that for you unless you tell what to do exactly when it compares elements with missing values.
In your situation you can provide a constant value to use for comparison whenever you compare vertices with missing values.
For example,
```
g.V().order().by(coalesce(values("name"), constant("")))
```

**Reply by li-boxuan** [reply]
That makes sense, but the behavior of JanusGraph and TinkerGraph diverge in this case. TinkerGraph simply ignores the vertex without that property.

**Reply by porunov** [reply]
I haven't checked how TinkerGraph does this, but I assume it would probably just move entities with missing values before or after entities which have some values. If that's the case, I guess we could have the same ordering behavior.

**Reply by joyemang33** [reply]
Thank you for your thorough investigation! As far as I know, in Tinker Graph, missing values are simply omitted from the entities. Another approach to ensure consistency could be to throw the same exception when such ordering is encountered in an empty traversal. I think that throwing the exception during execution and the inconsistency between the cases and the empty graph might pose challenges for downstream applications.
